### PR TITLE
Add config option `generated_file_header_size`

### LINKED ISSUE #5658: Option to configure how many lines down the @generated tag can be for format_generated_files?
In [Relay](https://github.com/facebook/relay) we have a large number of generated files which are required by our business to being with a specific copyright header that is slightly more than five lines. This means we can't take advantage of the format_generated_files feature. 

Example file: https://github.com/facebook/relay/blob/603bb4bc2927bda8b7f0dac1d86cbc5ecbf50202/compiler/crates/relay-lsp/tests/hover_test.rs

Would it be possible to make the 5 lines limit configurable?

### PR DESCRIPTION
**Pull request by captbaritone**
Fixes #5658

FYI @ytmimi

### CODE REVIEW

**Comment by captbaritone** [comment]
I wonder if it would be preferable to make `format_generated_files` a richer config option which the line number being a sub-option that can only be configured when the rule is enabled. As it stands, `generated_file_header_size` does not really make sense if `format_generated_files` is not enabled.

**Comment by ytmimi** [comment]
@captbaritone thanks for your first contribution to rustfmt 🎉

The tests are failing because we're missing an entry in [Configurations.md](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md) for the new config option. additionally it would be nice to [include some tests to validate the behavior of the new option](https://github.com/rust-lang/rustfmt/blob/master/Contributing.md#test-and-file-issues).

> I wonder if it would be preferable to make `format_generated_files` a richer config option which the line number being a sub-option that can only be configured when the rule is enabled.

Interesting idea! Are you suggesting the configuration should be something like this:

```
// Default. The file is always formatted
format_generated_files = true

// Disable formatting if the `@generated` marker is found within the first 5 lines of the file.
format_generated_files = false

// Disable formatting if the `@generated` marker is found within the first n lines of the file.
format_generated_files = <usize>
```

File: Configurations.md [hunk 1]
```diff
@@ -1050,15 +1050,23 @@ Max width for code snippets included in doc comments. Only used if [`format_code
...
 - **Possible values**: `true`, `false`
 - **Stable**: No (tracking issue: [#5080](https://github.com/rust-lang/rustfmt/issues/5080))
 
+## `generated_file_header_size`
+
+Number of lines to check for a `@generated` pragma header when `format_generated_files` is enabled. When `format_generated_files` is disabled, this option has no effect.
+
+- **Default value**: `5`
+- **Possible values**: any positive integer
+- **Stable**: No (tracking issue: [#5080](https://github.com/rust-lang/rustfmt/issues/5080))
```
Comment (captbaritone) [inline Configurations.md:1068]: Should these two config options be promoted to stable together? Or should I create a new issue to track promoting this item>

**Comment by captbaritone** [comment]
Hey, @ytmimi just following up to see if there's any feedback on this.

File: Configurations.md [hunk 2]
```diff
@@ -1050,15 +1050,23 @@ Max width for code snippets included in doc comments. Only used if [`format_code
...
 By default, generated files are reformatted, i. e. `@generated` marker is ignored.
 This option is currently ignored for stdin (`@generated` in stdin is ignored.)
 
 - **Default value**: `true`
 - **Possible values**: `true`, `false`
 - **Stable**: No (tracking issue: [#5080](https://github.com/rust-lang/rustfmt/issues/5080))
 
+## `generated_file_header_size`
+
+Number of lines to check for a `@generated` pragma header when `format_generated_files` is enabled. When `format_generated_files` is disabled, this option has no effect.
```
Comment (ytmimi) [inline Configurations.md:1064]: Let's reword this to be clearer about when this option has an effect:

Number of lines to check for a `@generated` pragma header when `format_generated_files` is false. When `format_generated_files` is true, this option has no effect.

File: Configurations.md (hunk 1 above)
Comment (ytmimi) [reply Configurations.md:1068]: I think it's would be fine to stabilize these at the same time.

File: Configurations.md [hunk 3]
```diff
@@ -1050,15 +1050,23 @@ Max width for code snippets included in doc comments. Only used if [`format_code
...
+
+- **Default value**: `5`
+- **Possible values**: any positive integer
+- **Stable**: No (tracking issue: [#5080](https://github.com/rust-lang/rustfmt/issues/5080))
+
```
Comment (ytmimi) [inline Configurations.md:1069]: Lets add a `See also [`format_generated_files`](#format_generated_files)` link here.

**Review by ytmimi — COMMENTED** [review]
Thanks for your work on this!

The implementation looks good, I'd just want us to make some minor tweaks to the documentation. Additionally can we add two more test cases:

1) Add a test case where `format_generated_files=true` and `generated_file_header_size` is set to some value where the `@generated` marker could be found. This would better test that when `format_generated_files=true` is set `generated_file_header_size` has no effect.
2) Add a test cast where `format_generated_files=false` and `generated_file_header_size` is set to some value where the `@generated` marker can't be found. This would show that even when `format_generated_files=false` we'll still format the file if we can't find the `@generaged` marker.

**Comment by calebcartwright** [comment]
Just want to add that I'm open to allowing the user to configure this value, but I haven't had a chance to consider this particular approach and I don't know if this is a significantly high enough priority for us to consider for release in the immediate future

**Comment by captbaritone** [comment]
@calebcartwright Thanks for the honesty. I'll wait for you to consider the approach more closely before I invest more time in this PR.

As it stands we are blocked from running rustftm from the CLI (needed to enforce formatting in CI) and must be careful to avoid saving generated files in our editor. Is there a workaround that you would recommend in the mean time?

**Comment by calebcartwright** [comment]
>  I'll wait for you to consider the approach more closely before I invest more time in this PR.

To be clear, we'll absolutely support adding an option that allows for this to be controlled, just need a chance to catch up on the back and forth within this thread as well as consider the specifics (as one initial reaction, something bothers me about the currently proposed option name as I don't think it's as self-explanatory as we'd like to see)

> As it stands we are blocked from running rustftm from the CLI (needed to enforce formatting in CI) and must be careful to avoid saving generated files in our editor. Is there a workaround that you would recommend in the mean time?
Merge state

Sorry not sure what you're asking. If you want to (temporarily) block folks from running rustfmt, intentionally or accidentally, the best route would be to enable [disable_all_formatting](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#disable_all_formatting). If you want to permit formatting in general but only skip generated files, then [ignore](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#ignore) would likely be a viable option, assuming you're on a nightly toolchain

**Comment by ytmimi** [comment]
closing since we merged https://github.com/rust-lang/rustfmt/pull/5993
